includes/Core/Util/Data_Request.php

```diff
@@ -0,0 +1,186 @@
+<?php
+/**
+ * Data_Request
+ *
+ * @package   Google\Site_Kit\Core\Util
+ * @copyright 2019 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Util;
+
+/**
+ * Class Data_Request
+ *
+ * @package Google\Site_Kit\Core\Util
+ */
+class Data_Request implements \ArrayAccess {
```

Comment:
Should we call this class `Request_Data` or something like that? The current naming is a bit strange because it's not a request, it's definition data for a request.

Most importantly, `create_data_request( Data_Request $data )` shows why the current naming is odd:
* The signature makes you wonder why you need to create a data request if you already pass a data request to the method.
* The parameter is called `$data` which is actually accurate.

I'm open to other naming suggestions, but I think the `Data_Request` name is confusing. I guess what it is is `Data_Request_Data`, but let's not do that. 😆

Reply:
I think `Request` is probably the simplest form and may be the best name for it.
- "Data" is always vague as to what it means.  
Perhaps `Google\Site_Kit\Core\REST_API` or a new `Google\Site_Kit\Data` would be a more suitable namespace?
- It also takes `$data` as a constructor argument (which is a bit confusing)
- Most of its arguments are about the request itself, data is just a bag of request parameters

Reply:
Maybe there's a more suitable term instead of `Data`. I think just calling it `Request` is still a bit odd, since `create_data_request()` generates an actual request from that. While this method is mostly an adapter to transform our object into a Google library-created guzzle request object, the naming as is still feels odd. `Data_Request_Definition`? (I think we should maintain the name of "data request" somehow because that has established itself as the name for this type of requests in the plugin, as great or not great that is)

Definitely +1 on moving that to the `Google\Site_Kit\Core\REST_API` namespace, let's do that.

Reply:
Not every request turns into a PSR7 Request during `create_data_request` but every request needs this context, very similar to a `WP_REST_Request`, so in that way `Request` is fitting, especially in the context of our `REST_API` namespace.